### PR TITLE
Fix GPT2 attention scaling ignored in SDPA/FlashAttention

### LINKED ISSUE #44380: GPT2 attention scaling config is ignored when using SDPA / FlashAttention backends
### System Info

None

### Who can help?

@ArthurZucker Hi, I'm new to LLMs and currently learning GPT2 model. I found that

The GPT2 attention configuration options:
	•	scale_attn_weights
	•	scale_attn_by_inverse_layer_idx

are respected in eager attention mode but silently ignored when using AttentionInterface backends such as "sdpa" or "flash_attention_2".

In eager mode:
the scaling logic is applied inside eager_attention_forward:
	•	division by sqrt(head_dim) if scale_attn_weights=True
	•	division by (layer_idx+1) if scale_attn_by_inverse_layer_idx=True

However, when using sdpa:
`torch._C._nn.scaled_dot_product_attention(query_states_3, key_states_3, value_states_3, attn_mask = attention_mask_1, dropout_p = 0.0, scale = None, is_causal = False)` which seems to ignore the above config.

I realize that the default configuration (`scale_attn_weights =True, scale_attn_by_inverse_layer_idx=False`) produces the same results, so I’m not sure whether this is intentional or should be considered a bug. 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

None

### Expected behavior

Different attention implementations should produce semantically equivalent results and respect model configuration parameters.

### PR DESCRIPTION
**Pull request by weiguangli-io**
# What does this PR do?

Fixes #44380

`GPT2Attention.forward()` did not pass the `scaling` parameter to `attention_interface`, causing `scale_attn_weights` and `scale_attn_by_inverse_layer_idx` config options to be silently ignored when using SDPA or FlashAttention backends.

The eager attention implementation (`eager_attention_forward`) reads these flags directly from the module and applies scaling correctly. However, SDPA and FlashAttention rely on the `scaling` parameter passed to the attention function call — which GPT2 never provided.

## Changes

1. **`modeling_gpt2.py`**: Compute `self.scaling` in `GPT2Attention.__init__` by combining `scale_attn_weights` (1/√d_k) and `scale_attn_by_inverse_layer_idx` (1/(layer_idx+1)), following the same pattern used by LLaMA and other models.
2. **`modeling_gpt2.py`**: Pass `scaling=self.scaling` to `attention_interface()` in `forward()`.
3. **`test_modeling_gpt2.py`**: Add `test_gpt2_sdpa_matches_eager_with_scaling_configs` that verifies SDPA and eager produce equivalent outputs when using non-default scaling configs.

## Behavior table (before → after)

| Config | Eager | SDPA before fix | SDPA after fix |
|--------|-------|-----------------|----------------|
| `scale_attn_weights=True` (default) | ÷√d_k | ÷√d_k (PyTorch default, coincidental match) | ÷√d_k ✓ |
| `scale_attn_weights=False` | No scaling | ÷√d_k (**wrong**) | No scaling ✓ |
| `scale_attn_by_inverse_layer_idx=True` | ÷(layer+1) | **Ignored** | ÷(layer+1) ✓ |

### CODE REVIEW

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -101,6 +101,12 @@ def __init__(self, config, is_cross_attention=False, layer_idx=None):
         self.layer_idx = layer_idx
         self.reorder_and_upcast_attn = config.reorder_and_upcast_attn
 
+        self.scaling = 1.0
+        if self.scale_attn_weights:
+            self.scaling = self.head_dim**-0.5
```
Comment (vasqu) [inline src/transformers/models/gpt2/modeling_gpt2.py:98]: I think we got lucky here because SDPA and FA will default to exactly this

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -101,6 +101,12 @@ def __init__(self, config, is_cross_attention=False, layer_idx=None):
...
+        if self.scale_attn_by_inverse_layer_idx:
+            self.scaling /= float(self.layer_idx + 1)
```
Comment (vasqu) [inline src/transformers/models/gpt2/modeling_gpt2.py:100]: This is what was silently ignored then

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (vasqu) [inline src/transformers/models/gpt2/modeling_gpt2.py:1]: Imo, we can just use `self.scaling` in the eager forward as well then and copy from bert or similar (meaning the eager forward). No need for extra treatment then

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 3]
```diff
@@ -246,6 +246,28 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
         )
         result.loss.backward()
 
+    def test_gpt2_sdpa_matches_eager_with_scaling_configs(self):
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py:250]: Could we also check FA

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 4]
```diff
@@ -246,6 +246,28 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        """Test that SDPA and eager produce equivalent outputs when scaling configs differ from defaults."""
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(scale_attn_by_inverse_layer_idx=True)
+        config, input_ids, token_type_ids, _, _, _, _ = config_and_inputs
+        config.scale_attn_weights = False
+        config.scale_attn_by_inverse_layer_idx = True
+
+        # Eager attention (known-correct reference)
+        config._attn_implementation = "eager"
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]: I'd rather use this after init, i.e. `model.set_attn_implementation("eager")`

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 5]
```diff
@@ -246,6 +246,28 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        model_eager = GPT2LMHeadModel(config).to(torch_device).eval()
+        with torch.no_grad():
+            output_eager = model_eager(input_ids, token_type_ids=token_type_ids).logits
+
+        # SDPA attention (was buggy: ignored scaling configs)
+        config._attn_implementation = "sdpa"
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]: Same here

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 6]
```diff
@@ -246,6 +246,28 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        model_sdpa = GPT2LMHeadModel(config).to(torch_device).eval()
+        model_sdpa.load_state_dict(model_eager.state_dict())
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]: We don't need this reloading stuff; we just switch up the flags with the `set_attn...`

**Review by vasqu — COMMENTED** [review]
Some smaller comments but we got lucky on the default case tbh. Very good finding

**Comment by weiguangli-io** [comment]
@vasqu Thanks for the review! Addressed all feedback:

1. **`self.scaling` in `_upcast_and_reordered_attn`**: Replaced the inline `scale_factor` computation with `self.scaling` in both GPT2 and DecisionTransformer. Now all attention paths consistently use `self.scaling` computed once in `__init__`.

2. **`set_attn_implementation` in tests**: Switched from model reloading with `config._attn_implementation` to `model.set_attn_implementation("eager")`/`"sdpa"` — single model, flag swap.

3. **FlashAttention2 test**: Added `test_gpt2_fa2_matches_eager_with_scaling_configs` gated behind `@require_torch_accelerator` and `@require_flash_attn`.

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (weiguangli-io) [reply src/transformers/models/gpt2/modeling_gpt2.py:1]: Thanks for the review! Addressed in the latest commit (04f9ba9f):

- `self.scaling` is computed once in `__init__` and used in both the `_upcast_and_reordered_attn` path (via `baddbmm alpha`) and the standard `attention_interface` path (via `scaling=self.scaling`).
- The old per-forward scale factor computation in `_upcast_and_reordered_attn` is removed.

So no extra treatment — just `self.scaling` everywhere, same pattern as bert.

File: tests/models/gpt2/test_modeling_gpt2.py (hunk 3 above)
Comment (weiguangli-io) [reply tests/models/gpt2/test_modeling_gpt2.py:250]: Added! The FA test is at `test_gpt2_fa2_matches_eager_with_scaling_configs` — both tests use `model.set_attn_implementation()` to switch backends without reloading.

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 7]
```diff
@@ -246,6 +246,52 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        model = GPT2LMHeadModel(config).to(torch_device).eval().to(torch.float16)
+        input_ids = input_ids.to(torch_device)
+        token_type_ids = token_type_ids.to(torch_device)
+
+        # Eager attention (known-correct reference)
+        model.set_attn_implementation("eager")
+        with torch.no_grad():
+            output_eager = model(input_ids, token_type_ids=token_type_ids).logits
+
+        # FlashAttention2
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]:
```suggestion
        # Flash Attention 2 (was buggy: ignored scaling configs)
```

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 8]
```diff
@@ -246,6 +246,52 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        model.set_attn_implementation("flash_attention_2")
+        with torch.no_grad():
+            output_fa2 = model(input_ids, token_type_ids=token_type_ids).logits
+
+        torch.testing.assert_close(output_eager, output_fa2, atol=5e-3, rtol=5e-3)
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]:
```suggestion
        torch.testing.assert_close(output_eager, output_fa2, atol=1e-2, rtol=1e-2)
```
Any reason for that atol/rtol? I remember FA being flaky at times so raising it would be personally preferred

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 9]
```diff
@@ -246,6 +246,52 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        model.set_attn_implementation("eager")
+        with torch.no_grad():
+            output_eager = model(input_ids, token_type_ids=token_type_ids).logits
+
+        # SDPA attention (was buggy: ignored scaling configs)
+        model.set_attn_implementation("sdpa")
+        with torch.no_grad():
+            output_sdpa = model(input_ids, token_type_ids=token_type_ids).logits
+
+        torch.testing.assert_close(output_eager, output_sdpa, atol=1e-5, rtol=1e-4)
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]:
```suggestion
        torch.testing.assert_close(output_eager, output_sdpa, atol=1e-4, rtol=1e-4)
```
same here, just a small raise to avoid flakiness

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (vasqu) [reply src/transformers/models/gpt2/modeling_gpt2.py:1]: Let's also address https://github.com/OiPunk/transformers/blob/04f9ba9ff0e3de8a8b21c801ba79509328ff14da/src/transformers/models/gpt2/modeling_gpt2.py#L54-L79 then

We can be closer to Bert since we unify into `self.scaling` instead https://github.com/OiPunk/transformers/blob/04f9ba9ff0e3de8a8b21c801ba79509328ff14da/src/transformers/models/bert/modeling_bert.py#L115-L140

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 10]
```diff
@@ -246,6 +246,52 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+
+    @require_torch_accelerator
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]:
```suggestion
    @require_torch_gpu
    @mark.flash_attn_test
```

**Review by vasqu — COMMENTED** [review]
Thanks for iterating, left another round of smaller comments to simplify a bit / follow some conventions

File: tests/models/gpt2/test_modeling_gpt2.py (hunk 10 above)
Comment (weiguangli-io) [reply tests/models/gpt2/test_modeling_gpt2.py]: Addressed in 802cd58:

- Refactored `eager_attention_forward` to accept `scaling` and `dropout` params, matching Bert's pattern — no more manual `module.scale_attn_weights` / `module.scale_attn_by_inverse_layer_idx` checks.
- Applied the reviewer's tolerance suggestions: SDPA `atol=1e-4`, FA2 `atol=1e-2`.
- Added `@require_torch_gpu` + `@pytest.mark.flash_attn_test` to the FA test.
- Fixed the FA comment text.
- Synced DecisionTransformer's `# Copied from` copies.

File: tests/models/gpt2/test_modeling_gpt2.py [hunk 11]
```diff
@@ -246,6 +247,53 @@ def test_gpt2_scale_attn_by_inverse_layer_idx(self):
...
+        with torch.no_grad():
+            output_sdpa = model(input_ids, token_type_ids=token_type_ids).logits
+
+        torch.testing.assert_close(output_eager, output_sdpa, atol=1e-4, rtol=1e-4)
+
+    @require_torch_gpu
+    @require_flash_attn
+    @pytest.mark.flash_attn_test
+    def test_gpt2_fa2_matches_eager_with_scaling_configs(self):
+        """Test that FlashAttention2 and eager produce equivalent outputs when scaling configs differ."""
```
Comment (vasqu) [inline tests/models/gpt2/test_modeling_gpt2.py]: Last nit: Let's add a small reference to the issue for both added tests

Checking with our slow CI in a second but looks good to me

**Review by vasqu — APPROVED** [review]
See my last comment but looks good

**Comment by vasqu** [comment]
run-slow: decision_transformer, gpt2

File: tests/models/gpt2/test_modeling_gpt2.py (hunk 11 above)
Comment (weiguangli-io) [reply tests/models/gpt2/test_modeling_gpt2.py]: Done in 309bb7f — added issue #44380 reference to both test docstrings.

**Comment by weiguangli-io** [comment]
Added issue #44380 reference to both test docstrings in 309bb7f. Ready for slow CI.

**Comment by vasqu** [comment]
This is super weird, no idea why the CI is failing here

**Comment by vasqu** [comment]
Thanks for contributing btw :hugs:
